api/pinsvcapi/pinsvcapi.go

```diff
@@ -0,0 +1,432 @@
+// Package pinsvcapi implements an IPFS Cluster API component which provides
+// an IPFS Pinning Services API to the cluster.
+//
+// The implented API is based on the common.API component (refer to module
+// description there). The only thing this module does is to provide route
+// handling for the otherwise common API component.
+package pinsvcapi
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/ipfs/go-cid"
+	types "github.com/ipfs/ipfs-cluster/api"
+	"github.com/ipfs/ipfs-cluster/api/common"
+	"github.com/ipfs/ipfs-cluster/api/pinsvcapi/pinsvc"
+	"github.com/ipfs/ipfs-cluster/state"
+	"go.uber.org/multierr"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	rpc "github.com/libp2p/go-libp2p-gorpc"
+)
+
+var (
+	logger    = logging.Logger("pinsvcapi")
+	apiLogger = logging.Logger("pinsvcapilog")
+)
+
+func trackerStatusToSvcStatus(st types.TrackerStatus) pinsvc.Status {
+	switch {
+	case st.Match(types.TrackerStatusError):
+		return pinsvc.StatusFailed
+	case st.Match(types.TrackerStatusPinQueued):
+		return pinsvc.StatusQueued
+	case st.Match(types.TrackerStatusPinning):
+		return pinsvc.StatusPinning
+	case st.Match(types.TrackerStatusPinned):
+		return pinsvc.StatusPinned
+	default:
+		return pinsvc.StatusUndefined
+	}
+}
+
+func svcStatusToTrackerStatus(st pinsvc.Status) types.TrackerStatus {
+	var tst types.TrackerStatus
+
+	if st.Match(pinsvc.StatusFailed) {
+		tst |= types.TrackerStatusError
+	}
+	if st.Match(pinsvc.StatusQueued) {
+		tst |= types.TrackerStatusPinQueued
+	}
+	if st.Match(pinsvc.StatusPinned) {
+		tst |= types.TrackerStatusPinned
+	}
+	if st.Match(pinsvc.StatusPinning) {
+		tst |= types.TrackerStatusPinning
+	}
+	return tst
+}
+
+func svcPinToClusterPin(p pinsvc.Pin) (*types.Pin, error) {
+	opts := types.PinOptions{
+		Name:     string(p.Name),
+		Origins:  p.Origins,
+		Metadata: p.Meta,
+		Mode:     types.PinModeRecursive,
+	}
+	c, err := cid.Decode(p.Cid)
+	if err != nil {
+		return nil, err
+	}
+	return types.PinWithOpts(c, opts), nil
+}
+
+func globalPinInfoToSvcPinStatus(
+	rID string,
+	gpi types.GlobalPinInfo,
+) pinsvc.PinStatus {
+
+	status := pinsvc.PinStatus{
+		RequestID: rID,
+	}
+
+	var statusMask types.TrackerStatus
+	for _, pinfo := range gpi.PeerMap {
+		statusMask |= pinfo.Status
+	}
+
+	status.Status = trackerStatusToSvcStatus(statusMask)
+	status.Pin = pinsvc.Pin{
+		Cid:     gpi.Cid.String(),
+		Name:    pinsvc.PinName(gpi.Name),
+		Origins: gpi.Origins,
+		Meta:    gpi.Metadata,
+	}
+
+	for _, pi := range gpi.PeerMap {
+		status.Delegates = append(status.Delegates, pi.IPFSAddresses...)
+		// Set created to the oldest known timestamp
+		if status.Created.IsZero() || pi.TS.Before(status.Created) {
+			status.Created = pi.TS
```

Comment:
On Status, for ongoing pin operations, it is set to the last phase change or to now() in case of error etc. For not ongoing operations as long as the pin exists, it is set to the pin's timestamp. Errors and unpinned items may come as time.now().

For StatusAll, it is set to the pin timestamp although overwritten by the optracker per the above (timestamp of the last phase change)...

We will add a Created field to pin status to avoid ambiguity.